apps/emqx_auth/src/emqx_auth_template.erl

```diff
@@ -24,61 +24,159 @@
     parse_sql/3,
+    cache_key_template/1,
+    cache_key/2,
+    cache_key/3,
+    placeholder_vars_from_str/1,
     render_deep_for_json/2,
     render_deep_for_url/2,
     render_deep_for_raw/2,
     render_str/2,
     render_urlencoded_str/2,
     render_sql_params/2,
     render_strict/2,
-    escape_disallowed_placeholders_str/2
+    escape_disallowed_placeholders_str/2,
+    rename_client_info_vars/1
 ]).
 
+-record(cache_key_template, {
+    id :: binary(),
+    vars :: emqx_template:t()
+}).
+
+-type var() :: emqx_template:varname() | {var_namespace, emqx_template:varname()}.
+-type allowed_vars() :: [var()].
+-type used_vars() :: [var()].
+-type cache_key_template() :: #cache_key_template{}.
+-type cache_key() :: fun(() -> term()).
+
 %%--------------------------------------------------------------------
 %% API
 %%--------------------------------------------------------------------
 
+-spec parse_deep(term(), allowed_vars()) -> {used_vars(), emqx_template:t()}.
 parse_deep(Template, AllowedVars) ->
     Result = emqx_template:parse_deep(Template),
     handle_disallowed_placeholders(Result, AllowedVars, {deep, Template}).
 
+-spec parse_str(unicode:chardata(), allowed_vars()) -> {used_vars(), emqx_template:t()}.
 parse_str(Template, AllowedVars) ->
     Result = emqx_template:parse(Template),
     handle_disallowed_placeholders(Result, AllowedVars, {string, Template}).
 
+-spec parse_sql(
+    emqx_template_sql:raw_statement_template(), emqx_template_sql:sql_parameters(), allowed_vars()
+) ->
+    {
+        used_vars(),
+        emqx_template_sql:statement(),
+        emqx_template_sql:row_template()
+    }.
 parse_sql(Template, ReplaceWith, AllowedVars) ->
     {Statement, Result} = emqx_template_sql:parse_prepstmt(
         Template,
         #{parameters => ReplaceWith, strip_double_quote => true}
     ),
-    {Statement, handle_disallowed_placeholders(Result, AllowedVars, {string, Template})}.
+    {UsedVars, TemplateWithAllowedVars} = handle_disallowed_placeholders(
+        Result, AllowedVars, {string, Template}
+    ),
+    {UsedVars, Statement, TemplateWithAllowedVars}.
 
+-spec cache_key_template(allowed_vars()) -> cache_key_template().
+cache_key_template(Vars) ->
+    #cache_key_template{
+        id = list_to_binary(emqx_utils:gen_id()),
+        vars = emqx_template:parse_deep(
+            lists:map(
+                fun(Var) ->
+                    list_to_binary("${" ++ Var ++ "}")
+                end,
+                Vars
+            )
+        )
+    }.
+
+-spec cache_key(map(), cache_key_template()) -> cache_key().
+cache_key(Values, CacheKeyTemplate) ->
+    cache_key(Values, CacheKeyTemplate, []).
+
+-spec cache_key(map(), cache_key_template(), list()) -> cache_key().
+cache_key(Values0, #cache_key_template{id = TemplateId, vars = KeyVars}, ExtraKeyParts) when
+    is_list(ExtraKeyParts)
+->
+    fun() ->
+        %% We hash the password because we do not want it to be stored as-is in the cache
```

Comment:
there can be secrets named other than `password` ? e.g. `secret_key` etc.?

Reply:
Only if put into client_attrs somehow. The other variables are predefined.